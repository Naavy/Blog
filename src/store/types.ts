export interface User {
  id: string;
  first_name: string;
  last_name: string;
  gender: string;
  dob: string;
  email: string;
  phone: string;
  website: string;
  address: string;
  status: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
    avatar: {
      href: string;
    }
  }

}

export interface Post {
  id: string;
  user_id: string;
  title: string;
  body: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
  }
}

export interface Comment {
  id: string;
  post_id: string;
  name: string;
  email: string;
  body: string;
  _links?: {
    self: {
      href: string;
    },
    edit: {
      href: string;
    }
  }
}

export interface AppState {
  users: User[];
  posts: Post[];
  comments: Comment[];
  selectedUser?: User;
  errors?: string;
  postsLoading: boolean;
  commentsLoading: boolean;
}

