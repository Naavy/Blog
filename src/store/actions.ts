import { Dispatch } from 'redux';
import { User, Post, Comment } from './types';

export const GET_USERS = 'GET_USERS';
export const GET_USERS_SUCCESS = 'GET_USERS_SUCCESS';
export const PATCH_USER = 'PATCH_USER';
export const PATCH_USER_SUCCESS = 'PATCH_USER_SUCCESS';
export const GET_POSTS = 'GET_POSTS';
export const GET_POSTS_SUCCESS = 'GET_POSTS_SUCCESS';
export const GET_COMMENTS = 'GET_COMMENTS';
export const GET_COMMENTS_SUCCESS = 'GET_COMMENTS_SUCCESS';
export const SELECT_USER = 'SELECT_USER';
export const ADD_COMMENT = 'ADD_COMMENT';
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS';
export const ADD_COMMENT_FAILURE = 'ADD_COMMENT_FAILURE';

const API = 'https://gorest.co.in/public-api';
const ACCESS_TOKEN = 'access-token=bfbcWHZ3j1wpEpMOlwuMGniVp7hQo6O3R4nt';


export const fetchUsers: any = () => {
  return (dispatch: Dispatch): void => {
    dispatch(getUsers());
    fetch(`${API}/users?${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getUsersSuccess(json.result)))
  };
};

export function getUsers() {
  return {
    type: GET_USERS
  }
}

export function getUsersSuccess(payload: User[]) {
  return {
    type: GET_USERS_SUCCESS,
    payload
  }
}

export function selectUser(payload: User) {
  return {
    type: SELECT_USER,
    payload
  }
}

export const changeUserName: any = (id: string, data: Record<string,string>) => {
  return (dispatch: Dispatch): void => {
    dispatch(patchUser(data));
    fetch(`${API}/users/${id}?${ACCESS_TOKEN}`,
      {
        method: 'PATCH',
        body: JSON.stringify(data),
        headers: {'Content-Type': 'application/json'}
      }
    )
      .then(response => response.json())
      .then(json => dispatch(patchUserSuccess(json.result)))
  };
};

export function patchUser(payload: Record<string,string>) {
  return {
    type: PATCH_USER,
    payload
  }
}

export function patchUserSuccess(payload: Record<string,string>) {
  return {
    type: PATCH_USER_SUCCESS,
    payload
  }
}

export const fetchPosts: any = (id: string) =>  {
  return (dispatch: Dispatch): void => {
    dispatch(getPosts());
    fetch(`${API}/posts?user_id=${id}&${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getPostsSuccess(json.result)))
  };
};

export function getPosts() {
  return {
    type: GET_POSTS
  }
}

export function getPostsSuccess(payload: Post[]) {
  return {
    type: GET_POSTS_SUCCESS,
    payload
  }
}

export const fetchComments: any = (id: string) =>  {
  return (dispatch: Dispatch): void => {
    dispatch(getComments());
    fetch(`${API}/comments?post_id=${id}&${ACCESS_TOKEN}`)
      .then(response => response.json())
      .then(json => dispatch(getCommentsSuccess(json.result)))
  };
};

export function getComments() {
  return {
    type: GET_COMMENTS
  }
}

export function getCommentsSuccess(payload: Comment[]) {
  return {
    type: GET_COMMENTS_SUCCESS,
    payload
  }
}

export const addComment: any = (comment: Comment) => {
  return (dispatch: Dispatch): void => {
    dispatch(addNewComment(comment));
    fetch(`${API}/comments?${ACCESS_TOKEN}`,
      {
        method: 'POST',
        body: JSON.stringify(comment),
        headers: {'Content-Type': 'application/json'}
      }
    )
      .then(response => response.json())
      .then(json => {
        if(json._meta.success) {
          dispatch(addNewCommentSuccess(json.result))
        } else { dispatch(addNewCommentFailure(json.result)) }
      })
  };
};

export function addNewComment(payload: Comment) {
  return {
    type: ADD_COMMENT,
    payload
  }
}

export function addNewCommentSuccess(payload: Comment) {
  return {
    type: ADD_COMMENT_SUCCESS,
    payload
  }
}

export function addNewCommentFailure(payload: any) {
  return {
    type: ADD_COMMENT_FAILURE,
    payload
  }
}
