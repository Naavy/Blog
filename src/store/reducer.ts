import { Reducer } from 'redux';
import { AppState } from './types';
import {
  GET_USERS_SUCCESS,
  SELECT_USER,
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_COMMENTS,
  GET_COMMENTS_SUCCESS,
  PATCH_USER_SUCCESS,
  ADD_COMMENT_SUCCESS,
  ADD_COMMENT_FAILURE
} from './actions';
import _ from 'lodash';

const initialState: AppState = {
  users: [],
  posts: [],
  comments: [],
  selectedUser: undefined,
  errors: undefined,
  postsLoading: false,
  commentsLoading: false
}

const reducer: Reducer<AppState> = (state = initialState, action) => {
  switch (action.type) {
    case GET_USERS_SUCCESS: {
      return {
        ...state,
        users: action.payload
      };
    }
    case SELECT_USER: {
      return {
        ...state,
        selectedUser: action.payload
      }
    }
    case GET_COMMENTS: {
      return {
        ...state,
        commentsLoading: true
      };
    }

    case GET_COMMENTS_SUCCESS: {
      return {
        ...state,
        comments:  _.uniqBy([...state.comments, ...action.payload], 'id'),
        commentsLoading: false
      };
    }

    case GET_POSTS: {
      return {
        ...state,
        postsLoading: true
      };
    }

    case GET_POSTS_SUCCESS: {
      return {
        ...state,
        posts: action.payload,
        comments: [],
        postsLoading: false
      };
    }
    case PATCH_USER_SUCCESS: {
      const newUsers = state.users.map((user) => {
        if(user.id === action.payload.id) {
          user = action.payload;
        }
        return user;
      })

      return {
        ...state,
        selectedUser: action.payload,
        users: newUsers
      };
    }

    case ADD_COMMENT_SUCCESS: {
      return {
        ...state,
        comments: [...state.comments, action.payload]
      };
    }

    case ADD_COMMENT_FAILURE: {
      return {
        ...state,
        errors: action.payload.message
      }
    }

    default: {
      return state;
    }
  }
};

export { reducer };
