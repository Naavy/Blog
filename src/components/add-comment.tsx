import React from 'react';
import { addComment } from '../store/actions';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { TextField, Button } from '@material-ui/core';
import { Post, Comment } from '../store/types';


interface State {
  newComment: string;
  name: string;
  email: string;
}

interface DispatchProps {
  addComment: (comment: Partial<Comment>) => any;
}

interface ComponentProps {
  post: Post
}

type Props = DispatchProps & ComponentProps;

export class AddComment extends React.Component<Props, State> {
  readonly state: State ={
    newComment: '',
    name: '',
    email: ''
  }

  sendForm(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    this.props.addComment({
      post_id: this.props.post.id,
      name: this.state.name,
      email: this.state.email,
      body: this.state.newComment
    });
  }

  render () {
    return (
      <form
        onSubmit={(event) => this.sendForm(event)}
        style={{ display: 'flex', flexDirection: 'column' }}
      >
        <TextField
          label="Your Name"
          value={this.state.name}
          onChange={(event) => this.setState({ name: event.target.value})}
        />
        <TextField
          label="Your e-mail"
          value={this.state.email}
          onChange={(event) => this.setState({ email: event.target.value})}
        />
        <TextField
          label="Your Comment..."
          multiline
          rowsMax={4}
          value={this.state.newComment}
          onChange={(event) => this.setState({ newComment: event.target.value })}
          variant="filled"
          size="small"
        />
        <Button
          variant="contained"
          type="submit"
          size="small"
          style={{ width: '50%', marginTop: '10px' }}
        >
          Add Comment
        </Button>
      </form>
    )
  }
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    addComment: (comment) => {
      return dispatch(addComment(comment))
    }
  }
}

export default connect(null, mapDispatchToProps)(AddComment);


