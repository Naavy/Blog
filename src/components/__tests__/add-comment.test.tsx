import React from 'react';
import { shallow } from 'enzyme';
import { AddComment } from '../add-comment';
import { Post } from '../../store/types';

const store = {
  newComment: '',
  name: '',
  email: '',
  addComment: jest.fn(),
  post: {} as Post
}

describe('AddComment', () => {
  it('should render', () => {
    // when
    const wrapper = shallow(<AddComment {...store} />);

    // then
    expect(wrapper).toMatchSnapshot();
  });

  it('should call addComment, when click "Add Comment" ', () => {
    // given
    const mockedAddComment = jest.fn().mockResolvedValue(null);
    const testStore = {
        ...store,
        addComment: mockedAddComment
    }

    const wrapper = shallow(<AddComment {...testStore} />);

    // when
    wrapper.find('form').simulate('submit', { preventDefault: () => null });

    // then
    expect(mockedAddComment).toHaveBeenCalled();
  });
});