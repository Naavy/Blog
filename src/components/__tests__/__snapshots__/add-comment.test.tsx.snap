// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComment should render 1`] = `ShallowWrapper {}`;
